diff --git a/dmenu.1 b/dmenu.1
index bbee17d..5717acd 100644
--- a/dmenu.1
+++ b/dmenu.1
@@ -48,6 +48,12 @@ X until stdin reaches end\-of\-file.
 .B \-i
 dmenu matches menu items case insensitively.
 .TP
+.B \-df
+dmenu matches the first choice by default.
+.TP
+.B \-S
+dmenu masks what is being typed.
+.TP
 .BI \-l " lines"
 dmenu lists items vertically, with the given number of lines.
 .TP
diff --git a/dmenu.c b/dmenu.c
index b56f3a8..35c43b9 100644
--- a/dmenu.c
+++ b/dmenu.c
@@ -55,6 +55,8 @@ static Item *prev, *curr, *next, *sel;
 static Window win;
 static XIC xic;
 static int mon = -1;
+static Bool defaultfirst = False;
+static Bool hidden = False;
 
 #include "config.h"
 
@@ -80,6 +82,10 @@ main(int argc, char *argv[]) {
 			fstrncmp = strncasecmp;
 			fstrstr = cistrstr;
 		}
+		else if(!strcmp(argv[i], "-df"))  /* use first line as default */
+			defaultfirst = True;
+		else if(!strcmp(argv[i], "-S"))  /* hidden input, e.g. password */
+			hidden = True;
 		else if(i+1 == argc)
 			usage();
 		/* these options take one argument */
@@ -162,6 +168,7 @@ void
 drawmenu(void) {
 	int curpos;
 	Item *item;
+	char hiddentext[sizeof text];
 
 	dc->x = 0;
 	dc->y = 0;
@@ -175,7 +182,11 @@ drawmenu(void) {
 	}
 	/* draw input field */
 	dc->w = (lines > 0 || !matches) ? mw - dc->x : inputw;
-	drawtext(dc, text, normcol);
+	if(hidden) {
+		memset(hiddentext, '*', strlen(text));
+		drawtext(dc, hiddentext, normcol);
+	} else
+		drawtext(dc, text, normcol);
 	if((curpos = textnw(dc, text, cursor) + dc->h/2 - 2) < dc->w)
 		drawrect(dc, curpos, 2, 1, dc->h - 4, True, FG(dc, normcol));
 
@@ -615,12 +626,18 @@ setup(void) {
 
 	XMapRaised(dc->dpy, win);
 	resizedc(dc, mw, mh);
+
+	if (defaultfirst && items) {
+		strncpy(text, items[0].text, BUFSIZ);
+		cursor = strlen(text);
+	}
+
 	drawmenu();
 }
 
 void
 usage(void) {
-	fputs("usage: dmenu [-b] [-f] [-i] [-l lines] [-p prompt] [-fn font] [-m monitor]\n"
+	fputs("usage: dmenu [-b] [-f] [-i] [-df] [-S] [-l lines] [-p prompt] [-fn font] [-m monitor]\n"
 	      "             [-nb color] [-nf color] [-sb color] [-sf color] [-v]\n", stderr);
 	exit(EXIT_FAILURE);
 }
